### PR TITLE
USDZExporter: Add hierarchy support, animation support, other improvements

### PR DESCRIPTION
**Pull request by hybridherbst**
**Description**

This PR aims to bring some of the improvements we did to our three.js fork of the USDZExporter back to the main repo.
For Needle Engine we're auto-converting not only animation, but also interactive behaviors on objects, so a number of changes were necessary to bring the USDZ output closer to glTF capabilities.

Related to #23415
Related to #21906

The `misc_exporter_usdz` example now has a complex animated model that also works in QuickLook / iOS AR: 

https://user-images.githubusercontent.com/2693840/236711976-da0857c9-6698-4cc1-9aae-0ae477c8f37e.mp4

**The main improvements are**
- hierarchy support (node hierarchies are exported as-is instead of all as flat hierarchy)
- animation support (enabled through hierarchy support) using timeSamples
- addition of a simple plugin system, which is used here for exporting animation data but can also be used for other cases (e.g. adding interactivity).
- addition of UV2 support (useful for occlusion textures)
- fixes to UsdTransform2d support (offset/repeat; no rotation for now) - QuickLook deviates from the spec, so there's two potential code paths in this PR, with QuickLook being the default since that's the main target for UDSZExporter if I'm not mistaken.
- addition of scale/bias support for textures, so colors don't need to be baked into textures anymore, they can be properly set.

**These limitations still exist:**
- no skinned mesh support (so the little wires on the bike don't get converted right now)
- no unlit conversion to lit + emissive (QuickLook doesn't support unlit)
- no vertex color conversion to textures (QuickLook doesn't support vertex colors)
- no blendshape support (QuickLook doesn't support blendshapes)

The main questions for this draft PR now are:
1. before spending more time on this PR, the main question: is the added complexity desired in three.js?
2. should the plugin system be made private and/or removed again so that just animation remains?

(Note: the Animation-related code is converted from TypeScript, there may still be a few traces of strange generated syntax left)

*This contribution is funded by [Needle](https://needle.tools)*

### CODE REVIEW

File: examples/jsm/exporters/USDZExporter.js [hunk 1]
```diff
@@ -617,53 +1239,330 @@ function buildVector2( vector ) {
...
+
+		if ( track.name.endsWith( 'scale' ) )
+			this.scale = track;
+
+	}
+
+	getFrames() {
+
+		var _a, _b, _c, _d, _e, _f, _g, _h, _j;
+		return Math.max( ( _c = ( _b = ( _a = this.pos ) === null || _a === void 0 ? void 0 : _a.times ) === null || _b === void 0 ? void 0 : _b.length ) !== null && _c !== void 0 ? _c : 0, ( _f = ( _e = ( _d = this.rot ) === null || _d === void 0 ? void 0 : _d.times ) === null || _e === void 0 ? void 0 : _e.length ) !== null && _f !== void 0 ? _f : 0, ( _j = ( _h = ( _g = this.scale ) === null || _g === void 0 ? void 0 : _g.times ) === null || _h === void 0 ? void 0 : _h.length ) !== null && _j !== void 0 ? _j : 0 );
```
Comment (LeviPesin) [inline examples/jsm/exporters/USDZExporter.js:1298]: What is this?

**Comment by mrdoob** [comment]
Big PRs with multiple changes are always difficult to work with...

I'm going to try to reimplement some of the things in the PR in new PRs.

File: examples/jsm/exporters/USDZExporter.js (hunk 1 above)
Comment (hybridherbst) [reply examples/jsm/exporters/USDZExporter.js:1298]: > (Note: the Animation-related code is converted from TypeScript, there may still be a few traces of strange generated syntax left)

I'm happy to clean this up if there's interest.

**Comment by mrdoob** [comment]
@hybridherbst How about doing a new PR with hierarchy support and then another PR with animation support?

**Comment by elalish** [comment]
This is excellent! I've been hoping for animation support for a long time. Any idea how much further work will be required for skeletal support?

**Comment by Mugen87** [comment]
>  How about doing a new PR with hierarchy support and then another PR with animation support?

Indeed! Let's split things up so it's easier to review and test.

**Comment by esquevin** [comment]
@hybridherbst would you still be interested in splitting this (old) MR? otherwise may I offer some assistance backporting your code to three.js?

**Comment by hybridherbst** [comment]
It’s supported in Needle for the time being. 

By now our USDZExporter doesn’t much resemble the original here, and supports hierarchy, animation, skinned meshes, display names, proper transforms, and even USDC. If there’s interest by maintainers I can look at making a new PR for _replacing_ the existing exporter.

**Comment by mrdoob** [comment]
@hybridherbst yea! let's try again 🙏

**Pull request closed without merging**